matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/create/CreateRoomFromLocalRoomTask.kt

```diff
@@ -0,0 +1,303 @@
+ */
+internal interface CreateRoomFromLocalRoomTask : Task<CreateRoomFromLocalRoomTask.Params, String> {
+    data class Params(val localRoomId: String)
+}
+
+internal class DefaultCreateRoomFromLocalRoomTask @Inject constructor(
+        @UserId private val userId: String,
+        @SessionDatabase private val monarchy: Monarchy,
+        private val createRoomTask: CreateRoomTask,
+        private val stateEventDataSource: StateEventDataSource,
+        private val clock: Clock,
+) : CreateRoomFromLocalRoomTask {
+
+    private val realmConfiguration
+        get() = monarchy.realmConfiguration
+
+    override suspend fun execute(params: CreateRoomFromLocalRoomTask.Params): String {
+        val replacementRoomId = stateEventDataSource.getStateEvent(params.localRoomId, EventType.STATE_ROOM_TOMBSTONE, QueryStringValue.IsEmpty)
+                ?.content?.toModel<RoomTombstoneContent>()
+                ?.replacementRoomId
+
+        if (replacementRoomId != null) {
+            return replacementRoomId
+        }
+
+        val createRoomParams = getCreateRoomParams(params)
+        val roomId = createRoomTask.execute(createRoomParams)
+
+        try {
+            // Wait for all the room events before triggering the replacement room
+            awaitNotEmptyResult(realmConfiguration, TimeUnit.MINUTES.toMillis(1L)) { realm ->
+                realm.where(RoomSummaryEntity::class.java)
+                        .equalTo(RoomSummaryEntityFields.ROOM_ID, roomId)
+                        .equalTo(RoomSummaryEntityFields.INVITED_MEMBERS_COUNT, createRoomParams.invitedUserIds.size.minus(1))
+            }
+            awaitNotEmptyResult(realmConfiguration, TimeUnit.MINUTES.toMillis(1L)) { realm ->
+                EventEntity.whereRoomId(realm, roomId)
+                        .equalTo(EventEntityFields.TYPE, EventType.STATE_ROOM_HISTORY_VISIBILITY)
+            }
+            if (createRoomParams.algorithm != null) {
+                awaitNotEmptyResult(realmConfiguration, TimeUnit.MINUTES.toMillis(1L)) { realm ->
+                    EventEntity.whereRoomId(realm, roomId)
+                            .equalTo(EventEntityFields.TYPE, EventType.STATE_ROOM_ENCRYPTION)
+                }
+            }
+        } catch (exception: TimeoutCancellationException) {
+            throw CreateRoomFailure.CreatedWithTimeout(roomId)
+        }
+
+        createTombstoneEvent(params, roomId)
+        return roomId
+    }
+
+    /* ==========================================================================================
+     * Private methods
+     * ========================================================================================== */
+
+    /**
+     * Retrieve the room configuration by parsing the state events related to the local room.
+     */
+    private suspend fun getCreateRoomParams(params: CreateRoomFromLocalRoomTask.Params): CreateRoomParams {
```

Comment:
To reduce the scope of the task, maybe we can move all code related to creating the `CreateRoomParams` into a dedicated class. I would be also easier to test. What do you think?

Reply:
I changed the internal implementation and removed this part